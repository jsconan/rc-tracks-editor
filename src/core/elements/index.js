/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

export { default as ArrowTip } from './ArrowTip.svelte';
export { default as CurvedArrow } from './CurvedArrow.svelte';
export { default as CurvedElement } from './CurvedElement.svelte';
export { default as CurvedElementEnlarged } from './CurvedElementEnlarged.svelte';
export { default as Shape } from './Shape.svelte';
export { default as Sketch } from './Sketch.svelte';
export { default as StraightArrow } from './StraightArrow.svelte';
export { default as StraightElement } from './StraightElement.svelte';
