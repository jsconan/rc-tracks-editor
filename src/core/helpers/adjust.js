/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

/**
 * The smallest value allowed to differ from a rounded value.
 * @type {number}
 * @private
 */
const EPSILON = 0.0000000000001;

/**
 * Adjusts a number so that it would not differ less than EPSILON from its rounded value.
 * @param {number} number - The number to adjust.
 * @returns {number} - The adjusted number.
 */
export default number => {
    const rounded = Math.round(number);
    if (Math.abs(rounded - number) <= EPSILON) {
        return rounded;
    }
    return number;
};
