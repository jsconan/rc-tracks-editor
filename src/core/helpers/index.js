/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

export { default as adjust } from './adjust.js';
export { default as alternate } from './alternate.js';
export { default as assign } from './assign.js';
export { default as defined } from './defined.js';
export { default as eachWord } from './eachWord.js';
export { default as getOwnPropertyDescriptors } from './getOwnPropertyDescriptors.js';
export { default as hasAPI } from './hasAPI.js';
export { default as mixin } from './mixin.js';
export { default as pick } from './pick.js';
export { default as uid } from './uid.js';
export { default as validateAPI } from './validateAPI.js';
export { default as validateCallback } from './validateCallback.js';
export { default as wait } from './wait.js';
export * from './angles.js';
export * from './compare.js';
export * from './svg.js';
export * from './transform.js';
