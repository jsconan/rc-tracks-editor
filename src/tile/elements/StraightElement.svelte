<script>
    // Licensed under GNU Public License version 3
    // Copyright (c) 2022 Jean-SÃ©bastien CONAN

    let cls = void 0;
    export { cls as class };
    export let style = void 0;
    export let x = 0;
    export let y = 0;
    export let width = 1;
    export let height = 1;
    export let fill = void 0;
    export let stroke = void 0;
    export let strokeWidth = void 0;
</script>

<rect class={cls} {style} {x} {y} {width} {height} {fill} {stroke} stroke-width={strokeWidth} />
