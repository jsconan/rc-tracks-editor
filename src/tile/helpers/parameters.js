/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import { CURVED_TILE_ENLARGED_TYPE, CURVED_TILE_TYPE, STRAIGHT_TILE_TYPE, validateType } from './tiles';

/**
 * Computes the parameters for rendering a tile at the expected position.
 * @type {object}
 * @private
 */
const tileParameters = {
    /**
     * Computes the parameters for rendering a curved tile at the expected position.
     * @param {CurvedTileModel} model - A reference to the tile model.
     * @param {number} tileX - The X-coordinate of the tile.
     * @param {number} tileY - The Y-coordinate of the tile.
     * @returns {object} - Returns a set parameters for rendering the tile at the expected position.
     */
    [CURVED_TILE_TYPE](model, tileX, tileY) {
        const specs = model.specs;
        const width = specs.width;
        const barrierWidth = specs.barrierWidth;
        const innerRadius = model.getInnerRadius();
        const outerRadius = model.getOuterRadius() - barrierWidth;
        const innerChunks = model.getInnerBarrierChunks();
        const outerChunks = model.getOuterBarrierChunks();
        const curveAngle = model.getCurveAngle();
        const curveCenter = model.getCurveCenter(tileX, tileY);

        const innerX = curveCenter.x + innerRadius;
        const innerY = curveCenter.y;
        const outerX = curveCenter.x + outerRadius;
        const outerY = curveCenter.y;

        const ground = {
            cx: curveCenter.x,
            cy: curveCenter.y,
            width,
            radius: innerRadius,
            angle: curveAngle,
            start: 0
        };
        const innerBarrier = {
            chunks: innerChunks,
            width: barrierWidth,
            radius: innerRadius,
            angle: curveAngle,
            left: innerX,
            top: innerY,
            shift: 0
        };
        const outerBarrier = {
            chunks: outerChunks,
            width: barrierWidth,
            radius: outerRadius,
            angle: curveAngle,
            left: outerX,
            top: outerY,
            shift: 1
        };

        return {
            ground,
            innerBarrier,
            outerBarrier
        };
    },

    /**
     * Computes the parameters for rendering a curved tile enlarged at the expected position.
     * @param {CurvedTileEnlargedModel} model - A reference to the tile model.
     * @param {number} tileX - The X-coordinate of the tile.
     * @param {number} tileY - The Y-coordinate of the tile.
     * @returns {object} - Returns a set parameters for rendering the tile at the expected position.
     */
    [CURVED_TILE_ENLARGED_TYPE](model, tileX, tileY) {
        const specs = model.specs;
        const width = specs.width;
        const barrierLength = specs.barrierLength;
        const barrierWidth = specs.barrierWidth;
        const side = model.getCurveSide();
        const innerRadius = model.getInnerRadius();
        const outerRadius = model.getOuterRadius() - barrierWidth;
        const sideChunks = model.getSideBarrierChunks();
        const innerChunks = model.getInnerBarrierChunks();
        const outerChunks = model.getOuterBarrierChunks();
        const curveAngle = model.getCurveAngle();
        const curveCenter = model.getCurveCenter(tileX, tileY);
        const outerBarrierPosition = width - barrierWidth;

        const { x: innerX, y: innerY } = curveCenter.addScalarX(innerRadius);
        const { x: outerX, y: outerY } = curveCenter.addCoord(
            side + outerRadius,
            innerRadius + outerBarrierPosition - outerRadius
        );
        const { x: horizontalX, y: horizontalY } = curveCenter.addScalarY(innerRadius + outerBarrierPosition);
        const { x: verticalX, y: verticalY } = curveCenter.addScalarX(innerRadius + outerBarrierPosition);

        const ground = {
            cx: curveCenter.x,
            cy: curveCenter.y,
            width,
            side,
            radius: innerRadius
        };
        const innerBarrier = {
            chunks: innerChunks,
            width: barrierWidth,
            radius: innerRadius,
            angle: curveAngle,
            left: innerX,
            top: innerY,
            shift: 0
        };
        const outerBarrier = {
            chunks: outerChunks,
            width: barrierWidth,
            radius: outerRadius,
            angle: curveAngle,
            left: outerX,
            top: outerY,
            shift: 1
        };
        const horizontalBarrier = {
            chunks: sideChunks,
            width: barrierWidth,
            length: barrierLength,
            left: horizontalX,
            top: horizontalY,
            shift: 0,
            vertical: false
        };
        const verticalBarrier = {
            chunks: sideChunks,
            width: barrierWidth,
            length: barrierLength,
            left: verticalX,
            top: verticalY,
            shift: 1,
            vertical: true
        };

        return {
            ground,
            innerBarrier,
            outerBarrier,
            horizontalBarrier,
            verticalBarrier
        };
    },

    /**
     * Computes the parameters for rendering a straight tile at the expected position.
     * @param {StraightTileModel} model - A reference to the tile model.
     * @param {number} tileX - The X-coordinate of the tile.
     * @param {number} tileY - The Y-coordinate of the tile.
     * @returns {object} - Returns a set parameters for rendering the tile at the expected position.
     */
    [STRAIGHT_TILE_TYPE](model, tileX, tileY) {
        const specs = model.specs;
        const barrierLength = specs.barrierLength;
        const barrierWidth = specs.barrierWidth;
        const width = model.width;
        const height = model.length;
        const chunks = model.getSideBarrierChunks();
        const innerRadius = model.getInnerRadius();
        const outerRadius = model.getOuterRadius();
        const curveCenter = model.getCurveCenter(tileX, tileY);

        const leftX = curveCenter.x + innerRadius;
        const leftY = curveCenter.y;
        const rightX = curveCenter.x + outerRadius - barrierWidth;
        const rightY = curveCenter.y;
        const vertical = true;

        const ground = {
            x: leftX,
            y: leftY,
            width,
            height
        };
        const leftBarrier = {
            chunks,
            width: barrierWidth,
            length: barrierLength,
            left: leftX,
            top: leftY,
            shift: 0,
            vertical
        };
        const rightBarrier = {
            chunks,
            width: barrierWidth,
            length: barrierLength,
            left: rightX,
            top: rightY,
            shift: 1,
            vertical
        };

        return { ground, leftBarrier, rightBarrier };
    }
};

/**
 * Computes the parameters for rendering a tile at the expected position.
 * @param {TileModel} model - A reference to the tile model.
 * @param {number} tileX - The X-coordinate of the tile.
 * @param {number} tileY - The Y-coordinate of the tile.
 * @returns {object} - Returns a set parameters for rendering the tile at the expected position.
 * @throws {TypeError} - If the given type is not valid.
 */

export function getTileParameters(model, tileX, tileY) {
    const { type } = model;
    validateType(type);
    return tileParameters[type](model, tileX, tileY);
}

/**
 * @typedef {import('../models').TileModel} TileModel
 */

/**
 * @typedef {import('../models').CurvedTileModel} CurvedTileModel
 */

/**
 * @typedef {import('../models').CurvedTileEnlargedModel} CurvedTileEnlargedModel
 */

/**
 * @typedef {import('../models').StraightTileModel} StraightTileModel
 */
