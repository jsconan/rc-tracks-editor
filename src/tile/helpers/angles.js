/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import { Vector2D } from '../../core/models';

const lastQuadrant = Vector2D.STRAIGHT_ANGLE + Vector2D.RIGHT_ANGLE;
const shiftAngle = Vector2D.RIGHT_ANGLE / 2;

/**
 * Returns the angle of the closest quadrant with respect to an angle.
 * @param {number} angle - The angle from which get the quadrant angle.
 * @returns {number} - Returns the angle of the closest quadrant.
 */
export const quadrantAngle = angle => {
    const edge = Vector2D.quadrant(angle + shiftAngle) * Vector2D.RIGHT_ANGLE;
    if (!edge && angle > lastQuadrant) {
        return Vector2D.CIRCLE;
    }
    return edge;
};

/**
 * Returns the angle of the closest quadrant included in a range of angles.
 * The range is not expected to be higher than 90 degrees wide.
 * @param {number} start - The start angle of the range.
 * @param {number} end - The end angle of the range.
 * @returns {number} - The quadrant angle contained by the range. It may be null.
 */
export const quadrantRange = (start, end) => {
    const startQuadrant = quadrantAngle(start);
    const endQuadrant = quadrantAngle(end);

    if (start < startQuadrant && startQuadrant < end) {
        return startQuadrant;
    }

    if (start < endQuadrant && endQuadrant < end) {
        return endQuadrant;
    }

    return null;
};
