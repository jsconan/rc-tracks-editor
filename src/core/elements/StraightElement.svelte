<script>
    // Licensed under GNU Public License version 3
    // Copyright (c) 2022 Jean-SÃ©bastien CONAN

    import { attributeList } from '../helpers';

    export let x = 0;
    export let y = 0;
    export let width = 1;
    export let height = 1;
    export let d = 0;
    export let fill = void 0;
    export let stroke = void 0;
    export let transform = void 0;
</script>

<rect
    x={x - d}
    y={y - d}
    width={width + d * 2}
    height={height + d * 2}
    {...attributeList(fill, 'fill')}
    {...attributeList(stroke, 'stroke')}
    {transform}
/>
