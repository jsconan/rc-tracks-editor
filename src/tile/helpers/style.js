/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import { alternate } from '../../core/helpers';

/**
 * The name of the style for the tile barriers.
 * @type {string}
 */
export const TILE_STYLE_BARRIER = 'barrier';

/**
 * The name of the style for the tile ground.
 * @type {string}
 */
export const TILE_STYLE_GROUND = 'ground';

/**
 * The name of the style for the selected tile.
 * @type {string}
 */
export const TILE_STYLE_SELECTED = 'selected';

/**
 * The name of the style for the hovered tile.
 * @type {string}
 */
export const TILE_STYLE_HOVERED = 'hovered';

/**
 * The name of the style for the focused tile.
 * @type {string}
 */
export const TILE_STYLE_FOCUSED = 'focused';

/**
 * A list of colors for the barriers.
 * @type {string[]}
 */
const barrierColors = ['even', 'odd'];

/**
 * Returns the presentation attributes for a barrier element at the given position.
 * @param {number} i - The position of the barrier element.
 * @returns {object} - The color for the barrier element.
 */
export const getTileStyleBarrier = i => ({
    fill: `var(--color-${alternate(i, barrierColors)})`
});

/**
 * Returns the presentation attributes for a tile ground.
 * @returns {object}
 */
export const getTileStyleGround = () => ({
    fill: 'var(--color-ground)',
    stroke: {
        stroke: 'var(--color-separator)',
        'stroke-width': 'var(--stroke-width-tile)'
    }
});

/**
 * Returns the presentation attributes of a tile overlay for a an selected state.
 * @returns {object}
 */
export const getTileStyleSelected = () => ({
    fill: 'var(--color-bg-selected)',
    stroke: {
        stroke: 'var(--color-fg-selected)',
        'stroke-width': 'var(--stroke-width-selected)'
    }
});

/**
 * Returns the presentation attributes of a tile overlay for a an hovered state.
 * @returns {object}
 */
export const getTileStyleHovered = () => ({
    fill: 'var(--color-bg-hovered)',
    stroke: {
        stroke: 'var(--color-fg-hovered)',
        'stroke-width': 'var(--stroke-width-hovered)'
    }
});

/**
 * Returns the presentation attributes of a tile overlay for a a focused state.
 * @returns {object}
 */
export const getTileStyleFocused = () => ({
    fill: 'var(--color-bg-focused)',
    stroke: {
        stroke: 'var(--color-fg-focused)',
        'stroke-width': 'var(--stroke-width-focused)',
        'stroke-dasharray': 'var(--stroke-dasharray-focused)'
    }
});

/**
 * Maps the styles to the helpers.
 * @type {object}
 * @private
 */
const styles = {
    [TILE_STYLE_BARRIER]: getTileStyleBarrier,
    [TILE_STYLE_GROUND]: getTileStyleGround,
    [TILE_STYLE_SELECTED]: getTileStyleSelected,
    [TILE_STYLE_HOVERED]: getTileStyleHovered,
    [TILE_STYLE_FOCUSED]: getTileStyleFocused
};

/**
 * Returns the presentation attributes by name.
 * @param {string} name - The name of the style to get.
 * @param  {...*} args - Additional parameters.
 * @returns {object}
 */
export const getTileStyle = (name, ...args) => {
    const style = styles[name];
    if ('function' === typeof style) {
        return style(...args);
    }

    return {};
};

/**
 * Extends a tile with the identified presentation attributes.
 * @param {string} name - The name of the style to set.
 * @param {object} tile - The tile to extend with the attributes.
 * @param  {...*} args - Additional parameters.
 * @returns {object}
 */
export const extendTileWithStyle = (name, tile, ...args) => {
    return { ...(tile || {}), ...getTileStyle(name, ...args) };
};
