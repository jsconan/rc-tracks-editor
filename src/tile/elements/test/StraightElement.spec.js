/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import { render } from '@testing-library/svelte';
import { wait } from '../../../core/helpers';
import StraightElement from '../StraightElement.svelte';

describe('StraightElement', () => {
    it('renders with default values', () => {
        const { container } = render(StraightElement);

        expect(container).toMatchSnapshot();
    });

    it('renders with the given parameters', () => {
        const props = {
            class: 'tile',
            style: 'fill: #444;',
            x: 100,
            y: 150,
            width: 80,
            height: 100
        };
        const { container } = render(StraightElement, { props });

        expect(container).toMatchSnapshot();
    });

    it.each([
        ['class', { class: 'curve' }],
        ['style', { style: 'fill: #888' }],
        ['x', { x: 40 }],
        ['y', { y: 40 }],
        ['width', { width: 40 }],
        ['height', { height: 40 }]
    ])('updates when the parameter %s is modified', async (title, update) => {
        const props = {
            class: 'tile',
            style: 'fill: #444;',
            x: 100,
            y: 150,
            width: 80,
            height: 100
        };
        const rendered = render(StraightElement, { props });

        return wait(10)
            .then(() => rendered.component.$set(update))
            .then(() => expect(rendered.container).toMatchSnapshot());
    });
});
