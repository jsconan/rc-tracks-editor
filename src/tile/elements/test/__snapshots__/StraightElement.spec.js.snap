// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StraightElement renders with default values 1`] = `
<body>
  <div>
    <rect
      height="1"
      width="1"
      x="0"
      y="0"
    />
  </div>
</body>
`;

exports[`StraightElement renders with the given parameters 1`] = `
<body>
  <div>
    <rect
      class="tile"
      fill="red"
      height="100"
      stroke="green"
      stroke-width="2"
      style="fill: #444;"
      transform="rotate(90)"
      width="80"
      x="100"
      y="150"
    />
  </div>
</body>
`;

exports[`StraightElement updates when the parameter class is modified 1`] = `
<body>
  <div>
    <rect
      class="curve"
      fill="red"
      height="100"
      stroke="green"
      stroke-width="2"
      style="fill: #444;"
      transform="rotate(90)"
      width="80"
      x="100"
      y="150"
    />
  </div>
</body>
`;

exports[`StraightElement updates when the parameter fill is modified 1`] = `
<body>
  <div>
    <rect
      class="tile"
      fill="blue"
      height="100"
      stroke="green"
      stroke-width="2"
      style="fill: #444;"
      transform="rotate(90)"
      width="80"
      x="100"
      y="150"
    />
  </div>
</body>
`;

exports[`StraightElement updates when the parameter height is modified 1`] = `
<body>
  <div>
    <rect
      class="tile"
      fill="red"
      height="40"
      stroke="green"
      stroke-width="2"
      style="fill: #444;"
      transform="rotate(90)"
      width="80"
      x="100"
      y="150"
    />
  </div>
</body>
`;

exports[`StraightElement updates when the parameter stroke is modified 1`] = `
<body>
  <div>
    <rect
      class="tile"
      fill="red"
      height="100"
      stroke="blue"
      stroke-width="2"
      style="fill: #444;"
      transform="rotate(90)"
      width="80"
      x="100"
      y="150"
    />
  </div>
</body>
`;

exports[`StraightElement updates when the parameter strokeWidth is modified 1`] = `
<body>
  <div>
    <rect
      class="tile"
      fill="red"
      height="100"
      stroke="green"
      stroke-width="3"
      style="fill: #444;"
      transform="rotate(90)"
      width="80"
      x="100"
      y="150"
    />
  </div>
</body>
`;

exports[`StraightElement updates when the parameter style is modified 1`] = `
<body>
  <div>
    <rect
      class="tile"
      fill="red"
      height="100"
      stroke="green"
      stroke-width="2"
      style="fill: #888"
      transform="rotate(90)"
      width="80"
      x="100"
      y="150"
    />
  </div>
</body>
`;

exports[`StraightElement updates when the parameter transform is modified 1`] = `
<body>
  <div>
    <rect
      class="tile"
      fill="red"
      height="100"
      stroke="green"
      stroke-width="2"
      style="fill: #444;"
      transform="rotate(180)"
      width="80"
      x="100"
      y="150"
    />
  </div>
</body>
`;

exports[`StraightElement updates when the parameter width is modified 1`] = `
<body>
  <div>
    <rect
      class="tile"
      fill="red"
      height="100"
      stroke="green"
      stroke-width="2"
      style="fill: #444;"
      transform="rotate(90)"
      width="40"
      x="100"
      y="150"
    />
  </div>
</body>
`;

exports[`StraightElement updates when the parameter x is modified 1`] = `
<body>
  <div>
    <rect
      class="tile"
      fill="red"
      height="100"
      stroke="green"
      stroke-width="2"
      style="fill: #444;"
      transform="rotate(90)"
      width="80"
      x="40"
      y="150"
    />
  </div>
</body>
`;

exports[`StraightElement updates when the parameter y is modified 1`] = `
<body>
  <div>
    <rect
      class="tile"
      fill="red"
      height="100"
      stroke="green"
      stroke-width="2"
      style="fill: #444;"
      transform="rotate(90)"
      width="80"
      x="100"
      y="40"
    />
  </div>
</body>
`;
